### ISSUE TITLE
pytorch latest update(1.4) broke MultiStepLR: wrong LR after step from _get_closed_form_lr

### ISSUE THREAD

**Issue by lironmo** [pytorch/pytorch#36913]
Hi,
the setting are simple,
using simple SDG optimizer with start lr of 0.05, and using MultiStepLR lr scheduler with milestone of 2,4.

code for reproduce: 
```python
from torch import optim
from torchvision.models import resnet50
from torch.optim import lr_scheduler


model = resnet50()
optimizer = optim.SGD(model.parameters(), lr=0.05)
ms_scheduler = lr_scheduler.MultiStepLR(optimizer=optimizer, milestones=[2,4])
lrs = [param_group['lr'] for param_group in optimizer.param_groups]
print(f"lr after lrs= {lrs}")
ms_scheduler.step(epoch=0)
lrs = [param_group['lr'] for param_group in optimizer.param_groups]
print(f"lr after lrs= {lrs}")
```

The bug happens if epoch is not None(i noted that that pytorch community tried to remove the option to pass epoch at step, but i will open a separate issue about that) at the __init__ of _LRScheduler there is a call to self._get_closed_form_lr() and reset the value of the optimizer. In my case the excepted value in epoch 0 is 0.05 and not  0.005(it's seems that he perform all the steps in the multistep in the init!!)

I see a code that handle this case when the epoch is None but buggy code for the other case.

I always pass the epoch when calling step( i have scheduler sequence wrapper that i wrote in order to change scheduler during training (most use cases are  warm up and tear down lr scheduling  ,but not only)

cc @vincentqb

**Comment by lironmo** [comment]
any update?

**Comment by vincentqb** [comment]
This works for me:

```python
model = resnet50()
optimizer = optim.SGD(model.parameters(), lr=0.05)
ms_scheduler = lr_scheduler.MultiStepLR(optimizer=optimizer, milestones=[2,4])
lrs = [param_group['lr'] for param_group in optimizer.param_groups]
print(f"lr at 0: {lrs}")

for i in range(10):
  lrs = [param_group['lr'] for param_group in optimizer.param_groups]
  print(f"lr at {i}: {lrs}")
   # train here
  ms_scheduler.step(epoch=None)
```

```
lr at 0: [0.05]
lr at 0: [0.05]
/private/home/vincentqb/anaconda3/envs/audio-built/lib/python3.7/site-packages/torch/optim/lr_scheduler.py:122: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  "https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate", UserWarning)
lr at 1: [0.05]
lr at 2: [0.005000000000000001]
lr at 3: [0.005000000000000001]
lr at 4: [0.0005000000000000001]
lr at 5: [0.0005000000000000001]
lr at 6: [0.0005000000000000001]
lr at 7: [0.0005000000000000001]
lr at 8: [0.0005000000000000001]
lr at 9: [0.0005000000000000001]
```

```python
model = resnet50()
optimizer = optim.SGD(model.parameters(), lr=0.05)
ms_scheduler = lr_scheduler.MultiStepLR(optimizer=optimizer, milestones=[2,4])
lrs = [param_group['lr'] for param_group in optimizer.param_groups]
print(f"lr at 0: {lrs}")

for i in range(1,10):
  lrs = [param_group['lr'] for param_group in optimizer.param_groups]
  print(f"lr at {i}: {lrs}")
   # train here
  ms_scheduler.step(epoch=i)
```

```
lr at 0: [0.05]
lr at 1: [0.05]
/private/home/vincentqb/anaconda3/envs/audio-built/lib/python3.7/site-packages/torch/optim/lr_scheduler.py:122: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  "https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate", UserWarning)
lr at 2: [0.0005000000000000001]
lr at 3: [0.0005000000000000001]
lr at 4: [0.0005000000000000001]
lr at 5: [0.0005000000000000001]
lr at 6: [0.0005000000000000001]
lr at 7: [0.0005000000000000001]
lr at 8: [0.0005000000000000001]
lr at 9: [0.0005000000000000001]
```

Does the code above resolve your issue?

Note that I am not calling `step(epoch=0)` here, since schedulers need to be called at the end of the loop, after the optimizer. Your issue is specifically with `step(epoch=0)` then, is that correct?

**Comment by vincentqb** [comment]
> I always pass the epoch when calling step( i have scheduler sequence wrapper that i wrote in order to change scheduler during training (most use cases are warm up and tear down lr scheduling ,but not only)

Can you describe what you are trying to do? There may be alternative approaches to using epoch as a parameter to step.

**Comment by lironmo** [comment]
@vincentqb thanks for your response.

in my case i have sequence of lr schedulers that being replace by predefined steps (that can be dynamic, i.e. increase over time ), for example from epoch 0-5 use warmup lr scheduler from 10-55 use cosine annealing scheduler from 55 to 100 use reduce on reduce on plateau.

i have a wrapper that switch between each scheduler and pass the epoch in each step, in order to make sure that each scheduler will be aware to the correct epoch.


in any case it's seems that this issue is fixed at pytorch 1.6(i tested it at pytorch 1.4 and got 0.005 but know i am getting the correct value 0.05 so we can close it.)
